Enable delegated proving in all transaction types

dagarcia7 opened the pull request:
#Summary 

Addresses https://github.com/0xPolygonMiden/miden-client/issues/736. Currently, the way that `new_mint_transaction`, `new_consume_transaction`, and `new_send_transaction` are implemented in the web client does not allow for these transactions to specify a remote prover if the user wants to enable delegated proving. Per the discussion in the issue, in this PR I have separated out the process of building the request for those types of transactions from the process of executing and submitting the transaction. Because `submitTransaction` allows for the user to specify a custom prover, this should solve the issue. 

Other changes included in this PR:
* Small refactor of the `/models/transaction_request.rs` file to break it up into smaller, more digestable files. 
* Fixed the existing web delegated prover testing infrastructure. There was an issue with the custom transaction test that was supposed to be testing passing a custom prover. Because it was missing a comma in the `it` statement, it was never actually being executed. Additionally, the `test:remote_prover` command specified in the `package.json` pointed to a different test altogether. Finally, the `Makefile` command was outdated for starting up a prover so that has been udpated as well.

dagarcia7 commented on crates/web-client/package.json:
```diff
@@ -19,8 +19,8 @@
     "build": "rimraf dist && rollup -c rollup.config.js && cpr js/types dist && node clean.js",
-    "test": "node --loader ts-node/esm ./node_modules/mocha/bin/mocha --file ./test/mocha.global.setup.mjs",
-    "test:remote_prover": "npm install && MIDEN_WEB_TESTING=true npm run build && cross-env REMOTE_PROVER=true node --loader ts-node/esm ./node_modules/mocha/bin/mocha --file ./test/mocha.global.setup.mjs -g new_mint_transactions tests",
+    "test": "node --loader ts-node/esm ./node_modules/mocha/bin/mocha --file ./test/mocha.global.setup.mjs --grep 'with remote prover' --invert",
```
This runs all tests that don't include `with remote prover` in their description. And the command below runs all the tests that include that description.

dagarcia7 commented on crates/web-client/src/new_transactions.rs:
```diff
@@ -197,14 +190,10 @@ impl WebClient {
+        Ok(consume_transaction_request.into())
     }
 
+    // TODO: Needs work
```
Will fix `swapTransaction` in a follow up PR

dagarcia7 commented on crates/web-client/test/new_transactions.test.ts:
```diff
@@ -552,21 +644,3 @@ describe("custom transaction with multiple output notes", () => {
-};
-
-describe("use custom transaction prover per request", () => {
-  it("custom transaction prover completes successfully"),
```
This was the test I thought was testing the delegated proving infrastructure. But this 

```
it("custom transaction prover completes successfully"), async () => { ... };
```
instead of this 

```
it("custom transaction prover completes successfully", async () => { ...  });
```

was causing this test to never be run correctly. Well, that and the fact that the `yarn:remote_prover` command specifically ran the `new_mint_transaction` test which showed up as a passing test 👍

dagarcia7 commented on crates/web-client/test/webClientTestUtils.ts:
There is some cleanup in general I could do across the `webClientTestUtils` file and the other testing files, but I'll reserve those for a follow up PR. Mainly, here there are two instances of `sendTransaction`... one here and one in the `new_transactions.test.ts` file so it's making it harder to maintain. but don't want to clutter this PR further from the actual changes needed.

tomyrd replied:
I also do some similar changes in #763 where I join those functions and also add some functions to execute the tx and instantly consume the output note.

tomyrd commented on crates/web-client/js/index.js:
```diff
@@ -340,49 +340,44 @@ export class WebClient {
     }
   }
 
-  async newConsumeTransaction(targetAccountId, noteId) {
```
Why do we remove the target account?

igamigo replied:
I think because now this would return a transaction request which can be executed by any account.
Although in this case, shouldn't the sender account be removed in the `newSendTransactionRequest()`?

dagarcia7 replied:
Yep, @igamigo is right! And to your question about why the sender account can't be removed from `newSendTransactionRequest()` is because in the send request, the sender and target accounts are needed to build a `PaymentTransactionData` object which is passed to the underlying `NativeTransactionRequestBuilder::pay_to_id` call. 

Same thing with `newMintTransaction`, the `targetAccountId` is still needed because `NativeTransactionRequestBuilder::mint_fungible_asset` needs it!

igamigo replied:
Oh, this is a good point. I believe we could actually remove it from `PaymentTransactionData`, but not for this PR!

tomyrd commented on crates/web-client/js/index.js:
```diff
@@ -340,49 +340,44 @@ export class WebClient {
-          targetAccountId,
-          noteId
-        );
+        return await this.wasmWebClient.newConsumeTransactionRequest(noteIds);
```
Should we remove the await here and in the other functions? Now that these functions are no longer async.

dagarcia7 replied:
Ah, yes I forgot to check these if statements. I removed the await from some other places, but missed these! Thanks for catching 👍

tomyrd commented on crates/web-client/src/new_transactions.rs:
I like these new changes, I considered doing something similar in #786 but decided to wait for a separate PR.

tomyrd commented on crates/web-client/test/new_transactions.test.ts:
```diff
@@ -67,33 +98,67 @@ export const sendTransaction = async (): Promise<SendTransactionResult> => {
+    if (_withRemoteProver && window.remoteProverUrl != null) {
+      await client.submitTransaction(
+        senderConsumeTransactionResult,
+        window.TransactionProver.newRemoteProver(window.remoteProverUrl)
```
Could we instantiate the remote prover once and then just use the same instance? Similar to what we do with the client

tomyrd reviewed (approved):
The changes look good! I'm all for cleaning up the tests a bit. I also did some refactoring in #763 but for specific tests where we mint and consume notes.We can merge this first and I can fix the conflicts.

igamigo commented on Makefile:
```diff
@@ -144,7 +144,7 @@ build-prover: update-prover-branch ## Update dependencies and build the prover b
 .PHONY: start-prover
 start-prover: ## Run prover. This requires the base repo to be present at `miden-base`
-	cd $(PROVER_DIR) && RUST_LOG=info cargo run --bin miden-proving-service $(PROVER_FEATURES_TESTING) --release --locked -- start-worker -p $(PROVER_PORT)
+	cd $(PROVER_DIR) && RUST_LOG=info cargo run --bin miden-proving-service $(PROVER_FEATURES_TESTING) --release --locked -- start-worker -p $(PROVER_PORT) --tx-prover --batch-prover
```
Is there a need for enabling the batch prover?

igamigo commented on crates/web-client/js/index.js:
```diff
@@ -340,49 +340,44 @@ export class WebClient {
-        MethodName.NEW_CONSUME_TRANSACTION,
-        serializedTargetAccountId,
-        noteId
+      const serializedTransactionRequestBytes = await this.callMethodWithWorker(
+        MethodName.NEW_CONSUME_TRANSACTION_REQUEST,
+        noteIds
       );
```
This now returns a `TransactionRequest`, correct? In this case, does the interface with the client need to go through the web worker or could it be done in-place somehow?

dagarcia7 replied:
Yeah you bring up a good point I also struggled with. The weird thing is that out of the three transaction request types (mint, consume, and send), consume has this call https://github.com/0xPolygonMiden/miden-client/blob/9370829463dc7cc32960342156032b13c9c80f94/crates/web-client/src/new_transactions.rs#L189 which forces it to be async. 

So I opted to keep the same pattern we had before across all three since I don't think there's harm in offloading the work to the web worker even if send and mint are synchronous. This way the users of the SDK await each request creation call. 

I'm on the fence about it, though so I can make send and mint completely sync and not use web workers and keep consume the same if you think that's best @igamigo!

igamigo replied:
I think it's fine. But in general, do we need to call `get_input_note_with_id_prefix()`? I think that's probably something that was leftover from the time in which the web SDK was more closely resembling the CLI rather than the client library. Perhaps it can be exposed separately and the user can decide to retrieve notes that way or something like that.
I usually think of `TransactionRequest` as just a description of what's going to be executed, consumed, etc. so it seemed like using the web worker seemed unnecessary but I don't have a strong opinion on this.

dagarcia7 replied:
Oh, good point. You're totally right. Now that we expose actual underlying types, I can easily either pass in a true list of `NoteId`s or recreate the `NoteId` object from the list of strings. So this is unnecessary. I'll fix this and now I think it totally makes sense to remove these altogether from the web workers! Great call 🙏

igamigo reviewed (commented):
Thanks for correcting the prover settings and adding tests! I went though the code and left some comments.
#786 will probably conflict with this PR a bit, but we can likely merge this first.

dagarcia7 commented:
@igamigo @tomyrd I think I addressed most of the feedback and should be ready to go soon, but running into issues with the proving service now. To double check, I reverted my makefile to checkout this commit on miden-base 911d713bbd5a9c41b8dcd576d1500e01af4896a2 and it worked locally. But fails with the latest commits.

tomyrd commented:
I was able to reproduce the CI error locally. I instantiated a prover using the make commands and ran `make integration-test-remote-prover-web-client`. The web errors aren't that descriptive (this gets fixed in #786) but it seems the error first comes up in the prover (by running it locally I have access to its logs):
```
ERROR proving_service:prove: miden-proving-service: error=status: Internal, message: "TransactionProgramExecutionFailed(NoMastForestWithProcedure { root_digest: RpoDigest([9713710825720099054, 3517411939058985379, 9226775740209413195, 9305841864214461414]) })", details: [], metadata: MetadataMap { headers: {} }
```

I think the problem is with the versions of miden-base used. The prover uses the latest `next` branch and the tests are using an old locked version. A `cargo update` will update the version but it will come with some errors that are being fixed in #793

dagarcia7 commented:
@tomyrd thanks for helping debug! I went ahead and rebased after the issue you linked got merged and ran a `cargo update` and pushed the changes to the `cargo.lock` file, but now getting different errors. I can revert the changes to the `cargo.lock` file, but then I'll see the prover issues as before. 

I think this is good to merge once these are resolved and we can turn these checks green so let me know if there's anything I can do or if there's any incoming work that I can keep a pulse on to retry!

tomyrd commented:
Just tested locally again. With the changes in #802  both the integration tests and the prover error were working. So they should be enough to fix the CI.

Edit: #802 got merged so the changes are now in `next`
